src/pal/src/misc/stackstring.cpp

```diff
@@ -0,0 +1,152 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information. 
+
+#include "pal/malloc.hpp"
+#include "pal/dbgmsg.h"
+
+SET_DEFAULT_DEBUG_CHANNEL(MISC);
+
+template <SIZE_T STACKCOUNT>
+class StackString
+{
+private:
+    WCHAR m_innerBuffer[STACKCOUNT + 1];
+    WCHAR * m_buffer;
+    SIZE_T m_count; // actual allocated count
+
+    void NullTerminate()
+    {
+        m_buffer[m_count] = W('\0');
+    }
+
+    void DeleteBuffer()
+    {
+        if (m_innerBuffer != m_buffer)
+            PAL_free(m_buffer);
+        else
+            m_buffer = NULL;
+
+        return;
+    }
+
+    void ReallocateBuffer(SIZE_T count)
+    {
+        // count is always > STACKCOUNT here.
+        WCHAR * newBuffer = (WCHAR *)PAL_malloc((count + 1) * sizeof(WCHAR));
+        if (NULL == newBuffer)
+        {
+            ERROR("malloc failed\n");
+            SetLastError(ERROR_NOT_ENOUGH_MEMORY);
+
+            DeleteBuffer();
+            m_count = 0;
+
+            return;
+        }
+
+        DeleteBuffer();
+        m_buffer = newBuffer;
+        m_count = count;
+
+        return;
+    }
+
+    void Resize(SIZE_T count)
+    {
+        if (NULL == m_buffer)
+        {
+            if (count > STACKCOUNT)
+            {
+                ReallocateBuffer(count);
+            }
+            else
+            {
+                m_buffer = m_innerBuffer;
+                m_count = count;
+            }
+        }
+        else if (m_innerBuffer == m_buffer)
+        {
+            if (count > STACKCOUNT)
+                ReallocateBuffer(count);
+            else
+                m_count = count;
+        }
+        else
+        {
+            ReallocateBuffer(count);
+        }
+
+        return;
+    }
+
+    StackString(const StackString &s)
+    {
+        Set(s);
+    }
+
+    ~StackString()
+    {
+        DeleteBuffer();
+    }
+
+public:
+    StackString(SIZE_T count)
```

Comment:
The constructor cannot report errors - I do not think this constructor should be here.